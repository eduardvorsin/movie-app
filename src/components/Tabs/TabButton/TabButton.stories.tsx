import type { Meta, StoryObj } from '@storybook/react';
import TabButton from './TabButton';

const meta: Meta<typeof TabButton> = {
	title: 'components/Tabs/TabButton',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for TabButton'
		},
		testId: {
			description: 'Id for testing the component',
		},
		isDisabled: {
			description: 'Allows you to make the component inactive',
		},
		isActive: {
			description: 'Shows whether the buttons and the associated tab are active or not',
		},
		onClick: {
			control: false,
			description: 'The callback that is called when clicked',
		},
		onKeyDown: {
			control: false,
			description: 'The callback that is called when the key is pressed',
		},
		ariaControls: {
			control: false,
			description: 'Allows you to set the id of the active tab for screen readers',
		},
		index: {
			control: false,
			description: 'Active tab index',
		},
	},
	component: TabButton,
	parameters: {
		docs: {
			description: {
				component: 'The TabButton component is an interactive button used in the Tabs component to select a specific tab or section on a web page. It provides user selection and navigation between different sections of content, which makes the interface more user-friendly and accessible to users.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof TabButton>;

export const Default: Story = {
	args: {
		id: 'tabs',
		label: 'tab-button',
		ariaControls: 'tab-panel-0'
	},
	parameters: {
		docs: {
			description: {
				story: 'Default state for TabButton',
			},
		},
	},
};

export const Active: Story = {
	args: {
		id: 'tabs',
		label: 'tab-button',
		ariaControls: 'tab-panel-0',
		isActive: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'The active state of the button',
			},
		},
	},
};

export const Disabled: Story = {
	args: {
		id: 'tabs',
		label: 'tab-button',
		ariaControls: 'tab-panel-0',
		isDisabled: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'Inactive state of the button',
			},
		},
	},
};

export const EventCallbacks: Story = {
	argTypes: {
		onClick: {
			action: 'Clicked',
		},
		onKeyDown: {
			action: 'Pressed',
		},
	},
	args: {
		id: 'tabs',
		label: 'tab-button',
		ariaControls: 'tab-panel-0',
		isActive: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'onClick and onKeyDown callbacks are called when the button is clicked and when the key is pressed',
			},
		},
	},
};