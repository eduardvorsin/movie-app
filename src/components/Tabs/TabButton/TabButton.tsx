import { GeneralProps } from '@/types/shared';
import { KeyboardEventHandler, MouseEventHandler } from 'react';

type Props = {
	id: string,
	label: string,
	isActive: boolean,
	ariaControls: string,
	onClick: MouseEventHandler<HTMLButtonElement>,
	onKeyDown: KeyboardEventHandler<HTMLButtonElement>,
	index: number,
	isDisabled?: boolean,
} & GeneralProps;

export default function TabButton({
	className,
	testId,
	id,
	isActive,
	label,
	index,
	isDisabled,
	ariaControls,
	onClick,
	onKeyDown,
	...props
}: Props) {
	const classes = [
		'text-100 font-medium p-1 rounded-[0.1875rem] text-center text-neutral-100 dark:text-dark-neutral-100 transition-colors duration-150',
		isActive ? 'bg-blue-700 dark:bg-blue-300 enabled:hover:bg-blue-800 enabled:active:bg-blue-900 dark:enabled:hover:bg-blue-200 dark:enabled:active:bg-blue-100' : 'bg-neutral-700 dark:bg-dark-neutral-700 enabled:hover:bg-neutral-800 enabled:active:bg-neutral-900 dark:enabled:hover:bg-dark-neutral-800 dark:enabled:active:bg-dark-neutral-900',
		isDisabled ? 'opacity-disabled cursor-not-allowed' : '',
		className,
	].join(' ');

	return (
		<button
			role='tab'
			id={id}
			className={classes}
			onClick={onClick}
			onKeyDown={onKeyDown}
			tabIndex={isActive ? 0 : -1}
			aria-selected={isActive}
			aria-controls={ariaControls}
			data-testid={testId}
			data-index={index}
			disabled={isDisabled}
			{...props}
		>
			{label}
		</button>
	);
}