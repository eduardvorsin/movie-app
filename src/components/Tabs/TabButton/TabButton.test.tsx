import { render, screen } from '@testing-library/react';
import TabButton from './TabButton';
import userEvent from '@testing-library/user-event';

describe('TabButton tests', () => {
	it('is rendered correctly', () => {
		render(
			<TabButton
				id='test-tab-button'
				label='test-tab-button'
				ariaControls='tab-panel-0'
				index={0}
				isActive={false}
				onClick={() => { }}
				onKeyDown={() => { }}
			/>
		);

		expect(screen.getByRole<HTMLButtonElement>('tab')).toBeInTheDocument();
	});

	it('when you click on the button, the mock function is triggered', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		render(
			<TabButton
				id='test-tab-button'
				label='test-tab-button'
				ariaControls='tab-panel-0'
				index={0}
				isActive={false}
				onClick={mockFn}
				onKeyDown={() => { }}
			/>
		);

		await user.click(screen.getByRole<HTMLButtonElement>('tab'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('when you press any key, the mock function is triggered', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		render(
			<TabButton
				id='test-tab-button'
				label='test-tab-button'
				ariaControls='tab-panel-0'
				index={0}
				isActive={true}
				onClick={() => { }}
				onKeyDown={mockFn}
			/>
		);

		await user.tab();
		await user.keyboard('[ArrowUp]');

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('if isDisabled is true, then the button becomes inactive', async () => {
		const user = userEvent.setup();
		const clickMockFn = jest.fn();
		const keyDownMockFn = jest.fn();
		render(
			<TabButton
				id='test-tab-button'
				label='test-tab-button'
				ariaControls='tab-panel-0'
				index={0}
				isActive={true}
				onClick={clickMockFn}
				onKeyDown={keyDownMockFn}
				isDisabled
			/>
		);
		await user.tab();
		await user.keyboard('[ArrowUp]');
		await user.click(screen.getByRole<HTMLButtonElement>('tab'));

		expect(clickMockFn).toHaveBeenCalledTimes(0);
		expect(keyDownMockFn).toHaveBeenCalledTimes(0);
	});

	it('is a basic snapshot', () => {
		render(
			<TabButton
				id='test-tab-button'
				label='test-tab-button'
				ariaControls='tab-panel-0'
				index={0}
				isActive={false}
				onClick={() => { }}
				onKeyDown={() => { }}
			/>
		);

		expect(screen.getByRole<HTMLButtonElement>('tab')).toMatchSnapshot();
	});

	it('snapshot with isDisabled set to true', () => {
		render(
			<TabButton
				id='test-tab-button'
				label='test-tab-button'
				ariaControls='tab-panel-0'
				index={0}
				isActive={false}
				isDisabled
				onClick={() => { }}
				onKeyDown={() => { }}
			/>
		);

		expect(screen.getByRole<HTMLButtonElement>('tab')).toMatchSnapshot();
	});

	it('snapshot with isActive equal to true', () => {
		render(
			<TabButton
				id='test-tab-button'
				label='test-tab-button'
				ariaControls='tab-panel-0'
				index={0}
				isActive={true}
				onClick={() => { }}
				onKeyDown={() => { }}
			/>
		);

		expect(screen.getByRole<HTMLButtonElement>('tab')).toMatchSnapshot();
	});
});