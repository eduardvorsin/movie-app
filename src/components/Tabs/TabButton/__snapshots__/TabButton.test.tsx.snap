// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabButton tests is a basic snapshot 1`] = `
<button
  aria-controls="tab-panel-0"
  aria-selected="false"
  class="text-100 font-medium p-1 rounded-[0.1875rem] text-center text-neutral-100 dark:text-dark-neutral-100 transition-colors duration-150 bg-neutral-700 dark:bg-dark-neutral-700 enabled:hover:bg-neutral-800 enabled:active:bg-neutral-900 dark:enabled:hover:bg-dark-neutral-800 dark:enabled:active:bg-dark-neutral-900  "
  data-index="0"
  id="test-tab-button"
  role="tab"
  tabindex="-1"
>
  test-tab-button
</button>
`;

exports[`TabButton tests snapshot with isActive equal to true 1`] = `
<button
  aria-controls="tab-panel-0"
  aria-selected="true"
  class="text-100 font-medium p-1 rounded-[0.1875rem] text-center text-neutral-100 dark:text-dark-neutral-100 transition-colors duration-150 bg-blue-700 dark:bg-blue-300 enabled:hover:bg-blue-800 enabled:active:bg-blue-900 dark:enabled:hover:bg-blue-200 dark:enabled:active:bg-blue-100  "
  data-index="0"
  id="test-tab-button"
  role="tab"
  tabindex="0"
>
  test-tab-button
</button>
`;

exports[`TabButton tests snapshot with isDisabled set to true 1`] = `
<button
  aria-controls="tab-panel-0"
  aria-selected="false"
  class="text-100 font-medium p-1 rounded-[0.1875rem] text-center text-neutral-100 dark:text-dark-neutral-100 transition-colors duration-150 bg-neutral-700 dark:bg-dark-neutral-700 enabled:hover:bg-neutral-800 enabled:active:bg-neutral-900 dark:enabled:hover:bg-dark-neutral-800 dark:enabled:active:bg-dark-neutral-900 opacity-disabled cursor-not-allowed "
  data-index="0"
  disabled=""
  id="test-tab-button"
  role="tab"
  tabindex="-1"
>
  test-tab-button
</button>
`;
